synapse/handlers/profile.py

```diff
@@ -38,11 +43,79 @@ def __init__(self, hs):
 
         self.user_directory_handler = hs.get_user_directory_handler()
 
+        self.http_client = hs.get_simple_http_client()
+
         if hs.config.worker_app is None:
             self.clock.looping_call(
                 self._update_remote_profile_cache, self.PROFILE_UPDATE_MS,
             )
 
+            if len(self.hs.config.replicate_user_profiles_to) > 0:
+                reactor.callWhenRunning(self._assign_profile_replication_batches)
+                reactor.callWhenRunning(self._replicate_profiles)
+                self.clock.looping_call(
+                    self._replicate_profiles, self.PROFILE_REPLICATE_INTERVAL
+                )
+
+    @defer.inlineCallbacks
+    def _assign_profile_replication_batches(self):
+        """If no profile replication has been done yet, allocate replication batch
+        numbers to each profile to start the replication process.
+        """
+        logger.info("Assigning profile batch numbers...")
+        total = 0
+        while True:
+            assigned = yield self.store.assign_profile_batch()
+            total += assigned
+            if assigned == 0:
+                break
+        logger.info("Assigned %d profile batch numbers", total)
+
+    @defer.inlineCallbacks
+    def _replicate_profiles(self):
```

Comment:
How does this handle overlapping calls? (e.g. two _replicate_profiles runs being kicked off by two users setting profile data at the same time, or colliding with the looping syncer)?

Reply:
They'll get different batch numbers depending on which one got serviced first, then probably the first call to _replicate_profiles will send both batches and the second will do nothing. If it collides with the looping syncer it'll depend which db txn gets done first.